File: chess_engine.py
import copy

COLOR_WHITE = 'w'
COLOR_BLACK = 'b'
PIECE_TYPES = {'K', 'Q', 'R', 'B', 'N', 'P'}


class Piece:
    __slots__ = ('type', 'color', 'has_moved')

    def __init__(self, type_, color, has_moved=False):
        self.type = type_
        self.color = color
        self.has_moved = has_moved

    def to_dict(self):
        return {'type': self.type, 'color': self.color, 'has_moved': self.has_moved}

    @staticmethod
    def from_dict(d):
        if d is None:
            return None
        return Piece(d['type'], d['color'], d.get('has_moved', False))

    def __repr__(self):
        return f"{self.color}{self.type}"


class Move:
    __slots__ = ('from_row', 'from_col', 'to_row', 'to_col',
                 'promotion', 'is_castle', 'is_en_passant', 'knighted', 'captured_piece')

    def __init__(self, from_row, from_col, to_row, to_col,
                 promotion=None, is_castle=False, is_en_passant=False, knighted=False):
        self.from_row = from_row
        self.from_col = from_col
        self.to_row = to_row
        self.to_col = to_col
        self.promotion = promotion
        self.is_castle = is_castle
        self.is_en_passant = is_en_passant
        self.knighted = knighted
        self.captured_piece = None

    def to_dict(self):
        return {
            'from_row': self.from_row, 'from_col': self.from_col,
            'to_row': self.to_row, 'to_col': self.to_col,
            'promotion': self.promotion, 'is_castle': self.is_castle,
            'is_en_passant': self.is_en_passant, 'knighted': self.knighted
        }

    def __eq__(self, other):
        return (self.from_row == other.from_row and self.from_col == other.from_col and
                self.to_row == other.to_row and self.to_col == other.to_col and
                self.promotion == other.promotion)

    def __repr__(self):
        return f"Move({self.from_row},{self.from_col}→{self.to_row},{self.to_col})"


class BoardState:
    def __init__(self):
        self.grid = [[None] * 8 for _ in range(8)]

    def setup_standard(self):
        back_row = ['R', 'N', 'B', 'Q', 'K', 'B', 'N', 'R']
        for col, pt in enumerate(back_row):
            self.grid[0][col] = Piece(pt, COLOR_BLACK)
            self.grid[7][col] = Piece(pt, COLOR_WHITE)
        for col in range(8):
            self.grid[1][col] = Piece('P', COLOR_BLACK)
            self.grid[6][col] = Piece('P', COLOR_WHITE)

    def get(self, row, col):
        if 0 <= row <= 7 and 0 <= col <= 7:
            return self.grid[row][col]
        return None

    def set(self, row, col, piece):
        self.grid[row][col] = piece

    def find_king(self, color):
        for r in range(8):
            for c in range(8):
                p = self.grid[r][c]
                if p and p.type == 'K' and p.color == color:
                    return (r, c)
        return None

    def clone(self):
        new_board = BoardState()
        for r in range(8):
            for c in range(8):
                p = self.grid[r][c]
                if p:
                    new_board.grid[r][c] = Piece(p.type, p.color, p.has_moved)
        return new_board

    def to_dict(self):
        return {
            'board': [
                [self.grid[r][c].to_dict() if self.grid[r][c] else None for c in range(8)]
                for r in range(8)
            ]
        }

    @staticmethod
    def from_dict(d):
        bs = BoardState()
        raw = d['board']
        for r in range(8):
            for c in range(8):
                bs.grid[r][c] = Piece.from_dict(raw[r][c])
        return bs


class ChessEngine:
    def get_raw_moves(self, board, row, col, effects):
        piece = board.get(row, col)
        if not piece:
            return []
        pt = piece.type
        color = piece.color

        # Rule: knights_rook — knights move like rooks
        if pt == 'N' and 'knights_rook' in effects:
            return self._moves_rook_like(board, row, col, color, effects)

        # Rule: pawns_knight — pawns move like knights
        if pt == 'P' and 'pawns_knight' in effects:
            return self._moves_knight_like(board, row, col, color, effects)

        # Rule: king_queen_move — kings also get queen slides
        dispatch = {
            'K': self._moves_king,
            'Q': self._moves_queen,
            'R': self._moves_rook,
            'B': self._moves_bishop,
            'N': self._moves_knight,
            'P': self._moves_pawn,
        }
        return dispatch[pt](board, row, col, color, effects)

    def get_legal_moves(self, board, row, col, color, effects, en_passant_target=None):
        raw = self.get_raw_moves(board, row, col, effects)
        legal = []
        for move in raw:
            if self._is_legal(board, move, color, effects, en_passant_target):
                legal.append(move)
        # Rule: forced_capture — must capture if any capture available
        if 'forced_capture' in effects:
            captures = [m for m in legal if m.captured_piece is not None or m.is_en_passant]
            if captures:
                return captures
        return legal

    def get_all_legal_moves(self, board, color, effects, en_passant_target=None):
        moves = []
        for r in range(8):
            for c in range(8):
                p = board.get(r, c)
                if p and p.color == color:
                    moves.extend(self.get_legal_moves(board, r, c, color, effects, en_passant_target))
        return moves

    def is_in_check(self, board, color, effects):
        king_pos = board.find_king(color)
        if not king_pos:
            return False
        opp = COLOR_BLACK if color == COLOR_WHITE else COLOR_WHITE
        for r in range(8):
            for c in range(8):
                p = board.get(r, c)
                if p and p.color == opp:
                    for mv in self.get_raw_moves(board, r, c, effects):
                        if (mv.to_row, mv.to_col) == king_pos:
                            return True
        return False

    def is_checkmate(self, board, color, effects, en_passant_target=None):
        if not self.is_in_check(board, color, effects):
            return False
        return len(self.get_all_legal_moves(board, color, effects, en_passant_target)) == 0

    def is_stalemate(self, board, color, effects, en_passant_target=None):
        if self.is_in_check(board, color, effects):
            return False
        return len(self.get_all_legal_moves(board, color, effects, en_passant_target)) == 0

    def _is_legal(self, board, move, color, effects, en_passant_target=None):
        sim = board.clone()
        self._apply_move_to_board(sim, move, en_passant_target)
        # After move, own king must not be in check
        return not self.is_in_check(sim, color, effects)

    def _apply_move_to_board(self, board, move, en_passant_target=None):
        piece = board.get(move.from_row, move.from_col)
        if not piece:
            return

        # En passant capture
        if move.is_en_passant and en_passant_target:
            ep_r, ep_c = en_passant_target
            # The captured pawn is on the same rank as the moving pawn, different from dest
            cap_row = move.from_row
            board.set(cap_row, ep_c, None)

        # Castling rook move
        if move.is_castle:
            if move.to_col == 6:  # kingside
                rook = board.get(move.from_row, 7)
                board.set(move.from_row, 5, rook)
                board.set(move.from_row, 7, None)
                if rook:
                    rook.has_moved = True
            else:  # queenside
                rook = board.get(move.from_row, 0)
                board.set(move.from_row, 3, rook)
                board.set(move.from_row, 0, None)
                if rook:
                    rook.has_moved = True

        board.set(move.from_row, move.from_col, None)
        new_piece = Piece(piece.type, piece.color, True)

        if move.knighted:
            new_piece.type = 'N'
        elif move.promotion:
            new_piece.type = move.promotion

        board.set(move.to_row, move.to_col, new_piece)
        move.captured_piece = board.get(move.to_row, move.to_col)

    # ── Move generators ──────────────────────────────────────────────────────

    def _moves_king(self, board, row, col, color, effects):
        moves = []
        for dr in (-1, 0, 1):
            for dc in (-1, 0, 1):
                if dr == 0 and dc == 0:
                    continue
                r, c = row + dr, col + dc
                if not (0 <= r <= 7 and 0 <= c <= 7):
                    continue
                target = board.get(r, c)
                if target and target.color == color and 'friendly_fire' not in effects:
                    continue
                if target and target.type == 'K':
                    continue
                moves.append(Move(row, col, r, c))

        # Rule: king_queen_move
        if 'king_queen_move' in effects:
            moves.extend(self._slide(board, row, col, color, effects,
                                     [(-1,0),(1,0),(0,-1),(0,1),(-1,-1),(-1,1),(1,-1),(1,1)]))

        # Castling (only if standard rules, king not moved)
        piece = board.get(row, col)
        if piece and not piece.has_moved and 'king_queen_move' not in effects:
            moves.extend(self._castle_moves(board, row, col, color, effects))

        return moves

    def _castle_moves(self, board, row, col, color, effects):
        moves = []
        # Kingside
        r7 = board.get(row, 7)
        if r7 and r7.type == 'R' and not r7.has_moved:
            if all(board.get(row, c) is None for c in (5, 6)):
                moves.append(Move(row, col, row, 6, is_castle=True))
        # Queenside
        r0 = board.get(row, 0)
        if r0 and r0.type == 'R' and not r0.has_moved:
            if all(board.get(row, c) is None for c in (1, 2, 3)):
                moves.append(Move(row, col, row, 2, is_castle=True))
        return moves

    def _moves_queen(self, board, row, col, color, effects):
        dirs = [(-1,0),(1,0),(0,-1),(0,1),(-1,-1),(-1,1),(1,-1),(1,1)]
        return self._slide(board, row, col, color, effects, dirs)

    def _moves_rook(self, board, row, col, color, effects):
        dirs = [(-1,0),(1,0),(0,-1),(0,1)]
        if 'rooks_diagonal' in effects:
            dirs += [(-1,-1),(-1,1),(1,-1),(1,1)]
        return self._slide(board, row, col, color, effects, dirs)

    def _moves_rook_like(self, board, row, col, color, effects):
        dirs = [(-1,0),(1,0),(0,-1),(0,1)]
        return self._slide(board, row, col, color, effects, dirs)

    def _moves_bishop(self, board, row, col, color, effects):
        dirs = [(-1,-1),(-1,1),(1,-1),(1,1)]
        if 'bishops_straight' in effects:
            dirs += [(-1,0),(1,0),(0,-1),(0,1)]
        return self._slide(board, row, col, color, effects, dirs)

    def _moves_knight(self, board, row, col, color, effects):
        offsets = [(-2,-1),(-2,1),(-1,-2),(-1,2),(1,-2),(1,2),(2,-1),(2,1)]
        moves = []
        for dr, dc in offsets:
            r, c = row + dr, col + dc
            if not (0 <= r <= 7 and 0 <= c <= 7):
                continue
            target = board.get(r, c)
            if target and target.color == color and 'friendly_fire' not in effects:
                continue
            if target and target.type == 'K':
                continue
            moves.append(Move(row, col, r, c))
        return moves

    def _moves_knight_like(self, board, row, col, color, effects):
        return self._moves_knight(board, row, col, color, effects)

    def _moves_pawn(self, board, row, col, color, effects):
        moves = []
        fwd = -1 if color == COLOR_WHITE else 1
        start_row = 6 if color == COLOR_WHITE else 1
        promo_row = 0 if color == COLOR_WHITE else 7
        mid_promo_row = 3 if color == COLOR_WHITE else 4  # for early_promote rule

        directions = [fwd]
        if 'pawns_backward' in effects:
            directions.append(-fwd)

        for direction in directions:
            # One step forward
            r1 = row + direction
            if 0 <= r1 <= 7 and board.get(r1, col) is None:
                promo = None
                if r1 == promo_row:
                    for pt in ('Q', 'R', 'B', 'N'):
                        moves.append(Move(row, col, r1, col, promotion=pt))
                elif 'early_promote' in effects and r1 == mid_promo_row:
                    moves.append(Move(row, col, r1, col, promotion='Q'))
                else:
                    moves.append(Move(row, col, r1, col))

                # Two steps from starting row (only forward direction)
                if direction == fwd and row == start_row:
                    r2 = row + 2 * direction
                    if 0 <= r2 <= 7 and board.get(r2, col) is None:
                        moves.append(Move(row, col, r2, col))

            # Diagonal captures
            for dc in (-1, 1):
                cr, cc = row + direction, col + dc
                if not (0 <= cr <= 7 and 0 <= cc <= 7):
                    continue
                target = board.get(cr, cc)
                if target and (target.color != color or 'friendly_fire' in effects) and target.type != 'K':
                    promo = None
                    if cr == promo_row:
                        for pt in ('Q', 'R', 'B', 'N'):
                            m = Move(row, col, cr, cc, promotion=pt)
                            m.captured_piece = target
                            moves.append(m)
                    else:
                        m = Move(row, col, cr, cc)
                        m.captured_piece = target
                        moves.append(m)

        return moves

    def get_en_passant_moves(self, board, row, col, color, en_passant_target):
        if not en_passant_target:
            return []
        ep_r, ep_c = en_passant_target
        fwd = -1 if color == COLOR_WHITE else 1
        piece = board.get(row, col)
        if not piece or piece.type != 'P':
            return []
        moves = []
        if row + fwd == ep_r and abs(col - ep_c) == 1:
            moves.append(Move(row, col, ep_r, ep_c, is_en_passant=True))
        return moves

    def _slide(self, board, row, col, color, effects, directions):
        moves = []
        can_jump = 'all_jump' in effects
        knighted_rule = 'knighted_on_jump' in effects

        for dr, dc in directions:
            r, c = row + dr, col + dc
            pieces_jumped = 0
            while 0 <= r <= 7 and 0 <= c <= 7:
                target = board.get(r, c)
                if target:
                    if can_jump:
                        if target.color == color and 'friendly_fire' not in effects:
                            pieces_jumped += 1
                            r += dr
                            c += dc
                            continue
                        if target.type == 'K':
                            break
                        knighted = knighted_rule and pieces_jumped >= 3
                        m = Move(row, col, r, c, knighted=knighted)
                        m.captured_piece = target
                        moves.append(m)
                        pieces_jumped += 1
                        r += dr
                        c += dc
                        continue
                    else:
                        # Standard: blocked by any piece
                        if target.color == color and 'friendly_fire' not in effects:
                            break
                        if target.type == 'K':
                            break
                        m = Move(row, col, r, c)
                        m.captured_piece = target
                        moves.append(m)
                        break
                else:
                    knighted = knighted_rule and can_jump and pieces_jumped >= 3
                    moves.append(Move(row, col, r, c, knighted=knighted))
                r += dr
                c += dc

        return moves

File: ChessGame/Engine/ChessEngine.cs
using ChessGame.Models;

namespace ChessGame.Engine;

public class ChessEngine
{
    public List<Move> GetRawMoves(BoardState board, int row, int col, HashSet<string> effects)
    {
        var piece = board.Get(row, col);
        if (piece == null) return [];

        // Rule overrides
        if (piece.Type == PieceType.N && effects.Contains("knights_rook"))
            return SlideRook(board, row, col, piece.Color, effects);
        if (piece.Type == PieceType.P && effects.Contains("pawns_knight"))
            return KnightOffsets(board, row, col, piece.Color, effects);

        return piece.Type switch
        {
            PieceType.K => KingMoves(board, row, col, piece.Color, effects),
            PieceType.Q => QueenMoves(board, row, col, piece.Color, effects),
            PieceType.R => RookMoves(board, row, col, piece.Color, effects),
            PieceType.B => BishopMoves(board, row, col, piece.Color, effects),
            PieceType.N => KnightOffsets(board, row, col, piece.Color, effects),
            PieceType.P => PawnMoves(board, row, col, piece.Color, effects),
            _ => []
        };
    }

    public List<Move> GetLegalMoves(BoardState board, int row, int col,
        PieceColor color, HashSet<string> effects, (int r, int c)? epTarget = null)
    {
        var raw = GetRawMoves(board, row, col, effects);

        // Add en passant for pawns
        if (board.Get(row, col)?.Type == PieceType.P)
            raw.AddRange(GetEnPassantMoves(board, row, col, color, epTarget));

        var legal = raw.Where(m => IsLegal(board, m, color, effects, epTarget)).ToList();

        // Forced capture rule
        if (effects.Contains("forced_capture"))
        {
            var captures = legal.Where(m => m.CapturedPiece != null || m.IsEnPassant).ToList();
            if (captures.Count > 0) return captures;
        }
        return legal;
    }

    public List<Move> GetAllLegalMoves(BoardState board, PieceColor color,
        HashSet<string> effects, (int r, int c)? epTarget = null)
    {
        var moves = new List<Move>();
        for (int r = 0; r < 8; r++)
            for (int c = 0; c < 8; c++)
                if (board.Get(r, c)?.Color == color)
                    moves.AddRange(GetLegalMoves(board, r, c, color, effects, epTarget));
        return moves;
    }

    public bool IsInCheck(BoardState board, PieceColor color, HashSet<string> effects)
    {
        var kingPos = board.FindKing(color);
        if (kingPos == null) return false;
        var opp = color == PieceColor.White ? PieceColor.Black : PieceColor.White;
        for (int r = 0; r < 8; r++)
            for (int c = 0; c < 8; c++)
                if (board.Get(r, c)?.Color == opp)
                    foreach (var mv in GetRawMoves(board, r, c, effects))
                        if (mv.ToRow == kingPos.Value.r && mv.ToCol == kingPos.Value.c)
                            return true;
        return false;
    }

    public bool IsCheckmate(BoardState board, PieceColor color,
        HashSet<string> effects, (int r, int c)? epTarget = null) =>
        IsInCheck(board, color, effects) &&
        GetAllLegalMoves(board, color, effects, epTarget).Count == 0;

    public bool IsStalemate(BoardState board, PieceColor color,
        HashSet<string> effects, (int r, int c)? epTarget = null) =>
        !IsInCheck(board, color, effects) &&
        GetAllLegalMoves(board, color, effects, epTarget).Count == 0;

    private bool IsLegal(BoardState board, Move move, PieceColor color,
        HashSet<string> effects, (int r, int c)? epTarget)
    {
        var sim = board.Clone();
        ApplyMoveToBoard(sim, move, epTarget);
        return !IsInCheck(sim, color, effects);
    }

    public void ApplyMoveToBoard(BoardState board, Move move, (int r, int c)? epTarget = null)
    {
        var piece = board.Get(move.FromRow, move.FromCol);
        if (piece == null) return;

        if (move.IsEnPassant && epTarget.HasValue)
            board.Set(move.FromRow, epTarget.Value.c, null);

        if (move.IsCastle)
        {
            int rookFromCol = move.ToCol == 6 ? 7 : 0;
            int rookToCol = move.ToCol == 6 ? 5 : 3;
            var rook = board.Get(move.FromRow, rookFromCol);
            board.Set(move.FromRow, rookToCol, rook);
            board.Set(move.FromRow, rookFromCol, null);
            if (rook != null) rook.HasMoved = true;
        }

        move.CapturedPiece = board.Get(move.ToRow, move.ToCol);
        board.Set(move.FromRow, move.FromCol, null);

        var newPiece = new Piece(piece.Type, piece.Color, true);
        if (move.IsKnighted) newPiece.Type = PieceType.N;
        else if (move.Promotion != null &&
                 Enum.TryParse<PieceType>(move.Promotion, out var pt))
            newPiece.Type = pt;

        board.Set(move.ToRow, move.ToCol, newPiece);
    }

    // ── Move generators ──────────────────────────────────────────────────────

    private List<Move> KingMoves(BoardState board, int r, int c,
        PieceColor color, HashSet<string> effects)
    {
        var moves = new List<Move>();
        for (int dr = -1; dr <= 1; dr++)
            for (int dc = -1; dc <= 1; dc++)
            {
                if (dr == 0 && dc == 0) continue;
                int nr = r + dr, nc = c + dc;
                if (nr < 0 || nr > 7 || nc < 0 || nc > 7) continue;
                var t = board.Get(nr, nc);
                if (t?.Type == PieceType.K) continue;
                if (t?.Color == color && !effects.Contains("friendly_fire")) continue;
                moves.Add(new Move(r, c, nr, nc));
            }

        if (effects.Contains("king_queen_move"))
            moves.AddRange(Slide(board, r, c, color, effects,
                [(-1,0),(1,0),(0,-1),(0,1),(-1,-1),(-1,1),(1,-1),(1,1)]));

        var piece = board.Get(r, c)!;
        if (!piece.HasMoved && !effects.Contains("king_queen_move"))
            moves.AddRange(CastleMoves(board, r, c, color));

        return moves;
    }

    private List<Move> CastleMoves(BoardState board, int r, int c, PieceColor color)
    {
        var moves = new List<Move>();
        var r7 = board.Get(r, 7);
        if (r7 is { Type: PieceType.R } && !r7.HasMoved &&
            board.Get(r, 5) == null && board.Get(r, 6) == null)
            moves.Add(new Move(r, c, r, 6, isCastle: true));
        var r0 = board.Get(r, 0);
        if (r0 is { Type: PieceType.R } && !r0.HasMoved &&
            board.Get(r, 1) == null && board.Get(r, 2) == null && board.Get(r, 3) == null)
            moves.Add(new Move(r, c, r, 2, isCastle: true));
        return moves;
    }

    private List<Move> QueenMoves(BoardState board, int r, int c,
        PieceColor color, HashSet<string> effects) =>
        Slide(board, r, c, color, effects,
            [(-1,0),(1,0),(0,-1),(0,1),(-1,-1),(-1,1),(1,-1),(1,1)]);

    private List<Move> RookMoves(BoardState board, int r, int c,
        PieceColor color, HashSet<string> effects)
    {
        var dirs = new List<(int, int)> { (-1,0),(1,0),(0,-1),(0,1) };
        if (effects.Contains("rooks_diagonal"))
            dirs.AddRange([(-1,-1),(-1,1),(1,-1),(1,1)]);
        return Slide(board, r, c, color, effects, dirs);
    }

    private List<Move> SlideRook(BoardState board, int r, int c,
        PieceColor color, HashSet<string> effects) =>
        Slide(board, r, c, color, effects, [(-1,0),(1,0),(0,-1),(0,1)]);

    private List<Move> BishopMoves(BoardState board, int r, int c,
        PieceColor color, HashSet<string> effects)
    {
        var dirs = new List<(int, int)> { (-1,-1),(-1,1),(1,-1),(1,1) };
        if (effects.Contains("bishops_straight"))
            dirs.AddRange([(-1,0),(1,0),(0,-1),(0,1)]);
        return Slide(board, r, c, color, effects, dirs);
    }

    private List<Move> KnightOffsets(BoardState board, int r, int c,
        PieceColor color, HashSet<string> effects)
    {
        var moves = new List<Move>();
        (int, int)[] offsets = [(-2,-1),(-2,1),(-1,-2),(-1,2),(1,-2),(1,2),(2,-1),(2,1)];
        foreach (var (dr, dc) in offsets)
        {
            int nr = r + dr, nc = c + dc;
            if (nr < 0 || nr > 7 || nc < 0 || nc > 7) continue;
            var t = board.Get(nr, nc);
            if (t?.Type == PieceType.K) continue;
            if (t?.Color == color && !effects.Contains("friendly_fire")) continue;
            moves.Add(new Move(r, c, nr, nc));
        }
        return moves;
    }

    private List<Move> PawnMoves(BoardState board, int r, int c,
        PieceColor color, HashSet<string> effects)
    {
        var moves = new List<Move>();
        int fwd = color == PieceColor.White ? -1 : 1;
        int startRow = color == PieceColor.White ? 6 : 1;
        int promoRow = color == PieceColor.White ? 0 : 7;
        int midPromoRow = color == PieceColor.White ? 3 : 4;

        var directions = new List<int> { fwd };
        if (effects.Contains("pawns_backward")) directions.Add(-fwd);

        foreach (int dir in directions)
        {
            int r1 = r + dir;
            if (r1 >= 0 && r1 <= 7 && board.Get(r1, c) == null)
            {
                if (r1 == promoRow)
                    foreach (var pt in new[] { "Q", "R", "B", "N" })
                        moves.Add(new Move(r, c, r1, c, promotion: pt));
                else if (effects.Contains("early_promote") && r1 == midPromoRow)
                    moves.Add(new Move(r, c, r1, c, promotion: "Q"));
                else
                    moves.Add(new Move(r, c, r1, c));

                if (dir == fwd && r == startRow)
                {
                    int r2 = r + 2 * dir;
                    if (r2 >= 0 && r2 <= 7 && board.Get(r2, c) == null)
                        moves.Add(new Move(r, c, r2, c));
                }
            }

            foreach (int dc in new[] { -1, 1 })
            {
                int cr = r + dir, cc = c + dc;
                if (cr < 0 || cr > 7 || cc < 0 || cc > 7) continue;
                var t = board.Get(cr, cc);
                if (t == null || t.Type == PieceType.K) continue;
                if (t.Color == color && !effects.Contains("friendly_fire")) continue;
                if (cr == promoRow)
                    foreach (var pt in new[] { "Q", "R", "B", "N" })
                    {
                        var m = new Move(r, c, cr, cc, promotion: pt);
                        m.CapturedPiece = t;
                        moves.Add(m);
                    }
                else
                {
                    var m = new Move(r, c, cr, cc);
                    m.CapturedPiece = t;
                    moves.Add(m);
                }
            }
        }
        return moves;
    }

    public List<Move> GetEnPassantMoves(BoardState board, int r, int c,
        PieceColor color, (int r, int c)? epTarget)
    {
        if (epTarget == null) return [];
        var piece = board.Get(r, c);
        if (piece?.Type != PieceType.P) return [];
        int fwd = color == PieceColor.White ? -1 : 1;
        if (r + fwd == epTarget.Value.r && Math.Abs(c - epTarget.Value.c) == 1)
            return [new Move(r, c, epTarget.Value.r, epTarget.Value.c, isEnPassant: true)];
        return [];
    }

    private List<Move> Slide(BoardState board, int r, int c, PieceColor color,
        HashSet<string> effects, IEnumerable<(int dr, int dc)> directions)
    {
        var moves = new List<Move>();
        bool canJump = effects.Contains("all_jump");
        bool knightedRule = effects.Contains("knighted_on_jump");

        foreach (var (dr, dc) in directions)
        {
            int nr = r + dr, nc = c + dc;
            int jumped = 0;
            while (nr >= 0 && nr <= 7 && nc >= 0 && nc <= 7)
            {
                var t = board.Get(nr, nc);
                if (t != null)
                {
                    if (canJump)
                    {
                        if (t.Color == color && !effects.Contains("friendly_fire"))
                        { jumped++; nr += dr; nc += dc; continue; }
                        if (t.Type == PieceType.K) break;
                        bool k = knightedRule && jumped >= 3;
                        var m = new Move(r, c, nr, nc, isKnighted: k);
                        m.CapturedPiece = t;
                        moves.Add(m);
                        jumped++;
                        nr += dr; nc += dc;
                        continue;
                    }
                    else
                    {
                        if (t.Color == color && !effects.Contains("friendly_fire")) break;
                        if (t.Type == PieceType.K) break;
                        var m = new Move(r, c, nr, nc);
                        m.CapturedPiece = t;
                        moves.Add(m);
                        break;
                    }
                }
                else
                {
                    bool k = knightedRule && canJump && jumped >= 3;
                    moves.Add(new Move(r, c, nr, nc, isKnighted: k));
                }
                nr += dr; nc += dc;
            }
        }
        return moves;
    }
}
